doc/iseltime.8
.TH ISELTIME 8 "Version 1.0: 13 Jul 2018"
.SH NAME
iseltime \- synchronize BIOS RTC and firmware SEL time from system time
.SH SYNOPSIS
.B "iseltime [-sx -N node -U user -P/-R pswd -EFJTVYZ]"

.SH DESCRIPTION
.I iseltime
is a program that allows synchronizing the hardware time values for 
RTC (BIOS Real-Time-Clock) and SEL (IPMI firmware SEL time) with 
the OS system time.  Run without options, iseltime shows each of these times 
to show whether they are the same or not.  

The OS system time should be in sync with the BIOS Real-Time-Clock.  
The OS system time may also be kept from drifting via an NTP server.  
Normally the BIOS will set the firmware SEL time from the RTC during boot, 
but running this command may be necessary periodically if the system does 
not reboot for a long time, for instance.  
Note that some applications may not handle setting the time backward well.
Note that  this command should not be run too frequently to avoid firmware 
timing issues.	It should not be set more than once per day.

.SH OPTIONS
Command line options are described below.
.IP "-s" 
Set the RTC and IPMI firmware SEL time to match the system time.  
This should not be performed more often than once a day.  

.IP "-x"
Causes eXtra debug messages to be displayed.
.IP "-N nodename"
Nodename or IP address of the remote target system.  If a nodename is
specified, IPMI LAN interface is used.  Otherwise the local system
management interface is used.
.IP "-U rmt_user"
Remote username for the nodename given.  The default is a null username.
.IP "-P/-R rmt_pswd"
Remote password for the nodename given.  The default is a null password.
.IP "-E"
Use the remote password from Environment variable IPMI_PASSWORD.
.IP "-F drv_t"
Force the driver type to one of the followng:  
imb, va, open, gnu, landesk, lan, lan2, lan2i, kcs, smb.
Note that lan2i means lan2 with intelplus.
The default is to detect any available driver type and use it.
.IP "-J"
Use the specified LanPlus cipher suite (0 thru 17): 0=none/none/none,
1=sha1/none/none, 2=sha1/sha1/none, 3=sha1/sha1/cbc128, 4=sha1/sha1/xrc4_128,
5=sha1/sha1/xrc4_40, 6=md5/none/none, ... 14=md5/md5/xrc4_40.
Default is 3.
.IP "-T"
Use a specified IPMI LAN Authentication Type: 0=None, 1=MD2, 2=MD5, 4=Straight Password, 5=OEM.
.IP "-V"
Use a specified IPMI LAN privilege level. 1=Callback level, 2=User level, 3=Operator level, 4=Administrator level (default), 5=OEM level.
.IP "-Y"
Yes, do prompt the user for the IPMI LAN remote password.
Alternatives for the password are -E or -P.
.IP "-Z"
Set the slave address for a local MC


.SH "SEE ALSO"
ipmiutil(8) ialarms(8) iconfig(8) icmd(8) idiscover(8) ievents(8) igetevent(8) ihealth(8) ilan(8) ireset(8) isel(8) isensor(8) iserial(8) isol(8) iwdt(8) 

.SH WARNINGS
See http://ipmiutil.sourceforge.net/ for the latest version of ipmiutil and any bug fix list. 

.SH COPYRIGHT
Copyright (C) 2009  Kontron America, Inc.
.PP
See the file COPYING in the distribution for more details
regarding redistribution.
.PP
This utility is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY.

.SH AUTHOR
.PP
Andy Cress <arcress at users.sourceforge.net>
.br



NEWS
IPMIUTIL NEWS 

In October 2001, this project started as part of the Carrier Grade Linux effort.
It was then known as 'panicsel' ("http://panicsel.sf.net") and included a kernel patch to write a Linux panic event to the IPMI firmware log, as well as utilities.  

In July 2002, this code was first included in MontaVista CGE Linux 2.1.
The panicsel functionality was included in OSDL CGL 1.0 and 2.0 requirements.
The kernel panic functionality was included in the OpenIPMI driver for 2.6 Linux kernels, and in the Intel IMB Linux driver v28 and greater.

A Linux rpm and a Windows setup package for ipmiutil is included on 
the platform CD with Intel carrier-grade servers.

The ipmiutil (or panicsel) rpm is known to be included in the following 
Linux distributions:
MontaVista CGE 2.1, 3.0, 3.1, 4.0, SuSE SLES9, SLES10, RedFlag 5.0

See http://ipmiutil.sourceforge.net for binaries, documentation, etc.

In Oct 2001 (0.8.1), started panicsel.
In Dec 2001 (1.0.0), released with Intel Telecom Linux Technology project
In Jul 2002 (1.2.1), released with MontaVista CGE 2.1
In Jan 2004 (1.4.1), added support for Windows 2000 & 2003
In Aug 2004 (1.4.8), the project was moved from panicsel.sf.net to ipmiutil.sf.net.  The new name more clearly reflects the purpose of the project in its current state.  
In Nov 2004 (1.5.0), support for IPMI LAN was added using FreeIPMI.
In May 2006 (1.7.0), built-in IPMI LAN support was added.
In Aug 2006 (1.7.5), driverless direct I/O support was added
In Jan 2007 (1.8.0), added ipmiutil meta-command, IPMI LAN 2.0 (lanplus) support
In Feb 2007 (1.9.2), added Linux Serial-Over-Lan console feature 
In Oct 2007 (2.0.0), added Windows SOL console and remote soft-shutdown features
In Mar 2008 (2.1.0), the Windows SOL client application was completed.
In Aug 2008 (2.2.0), iconfig function was added to save/restore the BMC configuration parameters
In Sep 2008 (2.2.1), support for the Microsoft IPMI driver was added.
In Nov 2008 (2.3.0), support for the Sun Solaris bmc driver was added.
In Mar 2009 (2.3.5), support for the FreeBSD OS was added.
In Oct 2009 (2.5.0), added ANSI VT100  emulation for Windows SOL console
In Feb 2010 (2.6.0), new subcommand naming scheme (i+subcmd)
In Apr 2010 (2.6.2), ifruset was added to set all FRU Product Area fields
In Jun 2010 (2.6.5), added picmg and firewall functionality
In Aug 2010 (2.6.8), Fedora 14 version of ipmiutil submitted
In Sep 2010 (2.6.9), added fwum/hpm functionality, and OEM Fujitsu sensor/sel 
In Sep 2010 (2.7.0), added sunoen, ekanalyzer, fru_picmg, oem_kontron
In Sep 2010 (2.7.1), added ppc64, mips64, sparc arch support
In Nov 2010 (2.7.2), added Debian package support
In Nov 2010 Fedora 14 released with ipmiutil-2.6.8
In Dec 2010 (2.7.3), added SuperMicro OEM functions
In Mar 2011 OpenSuSE 11.4 released with ipmiutil-2.7.2
In May 2011 (2.7.6), added VLAN support and WIN64 fixes
In Sep 2011 (2.7.8), added IPv6, systemd, Dell OEM functions, PICMG 2.3, tsol
In Oct 2011 (2.7.9), support Wind River and ARM cross-compile, Intel S2600
In Dec 2011 (2.8.0), added DCMI 1.1 commands, devel pkg, MSI package
In Feb 2012 (2.8.1), added Quanta S99Q functions, Dell fixes, Intel ME fixes
In Mar 2012 (2.8.2), added Intel Romley OEM support
In Jun 2012 (2.8.4), added support for HP-UX builds and DCMI 1.1 commands
In Oct 2012 (2.8.6), added pre-built Windows DLL for sample development apps
In Feb 2013 (2.8.8), added much more SuperMicro OEM support
In Apr 2013 (2.9.0), added iseltime to sync sys/RTC/SEL times
In May 2013 (2.9.1), added support for MacOSX client
In Dec 2016 (3.0.1), added support for Lenovo OEM sensors
In Feb 2017 (3.0.1), added support for ASUS OEM sensors
In Mar 2017 (3.0.3), added iuser to consolidate user functions



util/iseltime.c
/*
 * iseltime.c
 *
 * Author:  Andy Cress  arcress at users.sourceforge.net
 *
 * 05/11/09 Andy Cress v1.0 - created
 * 07/23/10 Andy Cress v1.1 - always show System time also
 * 08/20/10 Andy Cress v1.2 - show/set RTC time also if Linux
 * 03/16/17 Andy Cress 3.03 - show/set UTC offset also
 */
/*M*
Copyright (c) 2013, Andy Cress
All rights reserved.

Redistribution and use in source and binary forms, with or without 
modification, are permitted provided that the following conditions are met:

  a.. Redistributions of source code must retain the above copyright notice, 
      this list of conditions and the following disclaimer. 
  b.. Redistributions in binary form must reproduce the above copyright notice,
      this list of conditions and the following disclaimer in the documentation 
      and/or other materials provided with the distribution. 

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND 
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED 
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR 
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES 
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; 
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON 
ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT 
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS 
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *M*/
#ifdef WIN32
#include <windows.h>
#include <stdio.h>
#include "getopt.h"
#elif defined(EFI)
 // EFI: defined (EFI32) || defined (EFI64) || defined(EFIX64)
 #ifndef NULL
    #define NULL    0
 #endif
 #include <types.h>
 #include <libdbg.h>
 #include <unistd.h>
 #include <errno.h>
#elif defined(DOS)
 #include <dos.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
 #include "getopt.h"
#else
/* Linux, Solaris, BSD */
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#if defined(HPUX)
/* getopt is defined in stdio.h */
#elif defined(MACOS)
/* getopt is defined in unistd.h */
#include <unistd.h>
#else
#include <getopt.h>
#endif
#endif
#include <time.h>
#include <string.h>
#include "ipmicmd.h"
 
#define GET_SELTIME            0x48
#define SET_SELTIME            0x49
#define GET_SELUTC             0x5C
#define SET_SELUTC             0x5D
/*
 * Global variables 
 */
#ifdef METACOMMAND
extern char * progver;  /*from ipmiutil.c*/
static char * progname  = "ipmiutil seltime";
#else
static char * progver   = "3.08";
static char * progname  = "iseltime";
#endif
static char   fdebug    = 0;
static char   fset      = 0;
static uchar ipmi_maj = 0;
static uchar ipmi_min = 0;

static int get_sel_utc(uchar *rdata, int rlen)
{
	uchar idata[4];
	uchar ccode;
	int ret;
	ret = ipmi_cmdraw(GET_SELUTC, NETFN_STOR, BMC_SA, PUBLIC_BUS, BMC_LUN,
			  idata, 0, rdata, &rlen, &ccode, fdebug);
	if (ret == 0 && ccode != 0) ret = ccode;
	return(ret);
}  /*end get_sel_utc()*/

static int set_sel_utc(short offset)
{
	uchar idata[4];
	uchar rdata[16];
	int rlen = 8;
	int ret;
	uchar ccode;

	idata[0] = (uchar)(offset & 0x00ff);
	idata[1] = (uchar)((offset >> 8) & 0x00ff);
	ret = ipmi_cmdraw(SET_SELUTC, NETFN_STOR, BMC_SA, PUBLIC_BUS, BMC_LUN,
			  idata, 2, rdata, &rlen, &ccode, fdebug);
	if (ret == 0 && ccode != 0) ret = ccode;
	return(ret);
}  /*end set_sel_utc()*/

static int get_sel_time(uchar *rdata, int rlen)
{
	uchar idata[4];
	uchar ccode;
	int ret;
	ret = ipmi_cmdraw(GET_SELTIME, NETFN_STOR, BMC_SA, PUBLIC_BUS, BMC_LUN,
			  idata, 0, rdata, &rlen, &ccode, fdebug);
	if (ret == 0 && ccode != 0) ret = ccode;
	return(ret);
}  /*end get_sel_time()*/

static int set_sel_time(time_t newtime)
{
	uchar idata[4];
	uchar rdata[16];
	int rlen = 8;
	int ret;
	uchar ccode;

	idata[0] = (uchar)(newtime & 0x00ff);
	idata[1] = (uchar)((newtime >> 8) & 0x00ff);
	idata[2] = (uchar)((newtime >> 16) & 0x00ff);
	idata[3] = (uchar)((newtime >> 24) & 0x00ff);
	ret = ipmi_cmdraw(SET_SELTIME, NETFN_STOR, BMC_SA, PUBLIC_BUS, BMC_LUN,
			  idata, 4, rdata, &rlen, &ccode, fdebug);
	if (ret == 0 && ccode != 0) ret = ccode;
	return(ret);
}  /*end set_sel_time()*/

time_t utc2local(time_t t)
{
   struct tm * tm_tmp;
   int gt_year,gt_yday,gt_hour,lt_year,lt_yday,lt_hour;
   int delta_hour;
   time_t lt;

   //modify UTC time to local time expressed in number of seconds from 1/1/70 0:0:0 1970 GMT
   // check for dst?
   tm_tmp=gmtime(&t);
   gt_year=tm_tmp->tm_year;
   gt_yday=tm_tmp->tm_yday;
   gt_hour=tm_tmp->tm_hour;
   tm_tmp=localtime(&t);
   lt_year=tm_tmp->tm_year;
   lt_yday=tm_tmp->tm_yday;
   lt_hour=tm_tmp->tm_hour;
   delta_hour=lt_hour - gt_hour;
   if ( (lt_year > gt_year) || ((lt_year == gt_year) && (lt_yday > gt_yday)) )
           delta_hour += 24;
   if ( (lt_year < gt_year) || ((lt_year == gt_year) && (lt_yday < gt_yday)) )
           delta_hour -= 24;
   if (fdebug) printf("utc2local: delta_hour = %d\n",delta_hour);
   lt = t + (delta_hour * 60 * 60);
   return(lt);
}

#define TIMESTR_SZ  30
void show_time(time_t etime, int doutc)
{
	char buf[TIMESTR_SZ];
	int bufsz = TIMESTR_SZ;
	time_t t;

	strcpy(buf,"00/00/00 00:00:00");
	if (doutc) t = utc2local(etime);
	else t = etime;
	strftime(buf,bufsz, "%x %H:%M:%S", gmtime(&t)); /*or "%x %T"*/
	printf("%s\n",buf);
	return;
}

#ifdef METACOMMAND
int i_iseltime(int argc, char **argv)
#else
#ifdef WIN32
int __cdecl
#else
int
#endif
main(int argc, char **argv)
#endif
{
   int rv = 0;
   uchar devrec[20];
   uchar timebuf[4];
   time_t ltime1, ltime2, ltime3;
   short offset;
   int c;

#if defined (EFI)
        InitializeLib(_LIBC_EFIImageHandle, _LIBC_EFISystemTable);
#else
        printf("%s ver %s\n", progname,progver);
#endif

   while ( (c = getopt( argc, argv,"sT:V:J:EYF:P:N:R:U:x?")) != EOF ) 
      switch(c) {
          case 's': fset  = 1;   break;  /* read only */
          case 'x': fdebug = 1;     break;  /* debug messages */
          case 'N':    /* nodename */
          case 'U':    /* remote username */
          case 'P':    /* remote password */
          case 'R':    /* remote password */
          case 'E':    /* get password from IPMI_PASSWORD environment var */
          case 'F':    /* force driver type */
          case 'T':    /* auth type */
          case 'J':    /* cipher suite */ 
          case 'V':    /* priv level */
          case 'Y':    /* prompt for remote password */
                parse_lan_options(c,optarg,fdebug);
                break;
	  default:
                printf("Usage: %s [-sx -NUPRETVF]\n", progname);
                printf(" where -s      Set SEL time (usually once a day)\n");
                printf("       -x      show eXtra debug messages\n");
		print_lan_opt_usage(0);
                exit(1);
      }

   rv = ipmi_getdeviceid(devrec,16,fdebug);
   if (rv != 0) {
        show_outcome(progname,rv); 
	ipmi_close_();
	exit(rv);
   } else {
      ipmi_maj = devrec[4] & 0x0f;
      ipmi_min = devrec[4] >> 4;
#ifndef EFI
      printf("-- BMC version %x.%x, IPMI version %d.%d \n",
             devrec[2],  devrec[3], ipmi_maj, ipmi_min);
#endif
   }

   offset = 0;
   rv = get_sel_utc((uchar *)&offset,2);
   if (fdebug) printf("get_sel_utc: ret=%x offset=%d\n",rv,offset);
   if (rv == 0) {
      /* TODO: get system UTC offset */
      rv = set_sel_utc(offset);
      printf("set_sel_utc(%d): ret=%x\n",offset,rv);
   }
   rv = get_sel_time(&timebuf[0],4);
   if (rv != 0) {
	  printf("get_sel_time error: ret = %x\n",rv);
	  ipmi_close_();
	  exit(1);
   }
   time(&ltime2);
   printf("Current System time: "); show_time(ltime2,1);
   ltime1 = timebuf[0] + (timebuf[1] << 8) + (timebuf[2] << 16) + 
		(timebuf[3] << 24);
   printf("Current SEL time:    "); show_time(ltime1,0);
   
   // if (fdebug) ltime3 = utc2local(ltime1); 

   if (fset == 1) {
      /* get current time */
      time(&ltime2);
      rv = set_sel_time(ltime2);
      printf("Setting SEL time to System Time: ret = %x\n",rv); 
      if (rv != 0) printf("set_sel_time error: ret = %x\n",rv);
      else {  /*successful*/
	rv = get_sel_time(timebuf,8);
	if (rv != 0) printf("get_sel_time error: ret = %x\n",rv);
	else {
	   ltime3 = timebuf[0] + (timebuf[1] << 8) + (timebuf[2] << 16) + 
			(timebuf[3] << 24);
	   printf("New SEL time:        "); show_time(ltime3,0);
	}
      }
   } 
#ifdef LINUX
   if (is_remote() == 0) {
      c = system("echo \"Current RTC time:    `hwclock`\"");
      if (fset == 1) {
         c = system("hwclock --systohc");
	 printf("Copying System Time to RTC: ret = %d\n",c); 
      }
   }
#endif
   ipmi_close_();
   show_outcome(progname,rv); 
   exit (rv);
}  /* end main()*/

/* end iseltime.c */
